src/web/rustdoc.rs

```diff
@@ -56,171 +51,205 @@ static DOC_RUST_LANG_ORG_REDIRECTS: Lazy<HashMap<&str, &str>> = Lazy::new(|| {
+            // redirect all ico requests
+            // originally from:
+            // https://github.com/rust-lang/docs.rs/commit/f3848a34c391841a2516a9e6ad1f80f6f490c6d0
+            return Ok(axum_cached_redirect(
+                "/-/static/favicon.ico",
+                CachePolicy::ForeverInCdnAndBrowser,
+            )?
+            .into_response());
+        }
+    }
 
-    // this handler should never called without crate pattern
-    let crate_name = cexpect!(req, router.find("crate"));
-    let crate_name = percent_decode(crate_name.as_bytes())
-        .decode_utf8()
-        .unwrap_or_else(|_| crate_name.into());
-    let (mut crate_name, path_in_crate) = match crate_name.split_once("::") {
+    let (mut crate_name, path_in_crate) = match params.name.split_once("::") {
         Some((krate, path)) => (krate.to_string(), Some(path.to_string())),
-        None => (crate_name.to_string(), None),
+        None => (params.name.to_string(), None),
     };
 
     if let Some(inner_path) = DOC_RUST_LANG_ORG_REDIRECTS.get(crate_name.as_str()) {
-        let url = format!("https://doc.rust-lang.org/{inner_path}/");
-        return redirect_to_doc(
-            req,
-            url,
+        return Ok(redirect_to_doc(
+            &query_pairs,
+            format!("https://doc.rust-lang.org/{inner_path}/"),
             CachePolicy::ForeverInCdnAndStaleInBrowser,
             path_in_crate.as_deref(),
-        );
+        )?
+        .into_response());
     }
 
-    let req_version = router.find("version");
-    let mut target = router.find("target");
-
     // it doesn't matter if the version that was given was exact or not, since we're redirecting
     // anyway
     rendering_time.step("match version");
-    let v = match_version(&mut conn, &crate_name, req_version)?;
+    let v = match_version_axum(&pool, &crate_name, params.version.as_deref()).await?;
+    trace!(?v, "matched version");
     if let Some(new_name) = v.corrected_name {
         // `match_version` checked against -/_ typos, so if we have a name here we should
         // use that instead
         crate_name = new_name;
     }
     let (mut version, id) = v.version.into_parts();
 
-    if let None | Some("latest") = req_version {
+    // we might get requests to crate-specific JS files here.
+    if let Some(ref target) = params.target {
+        if target.ends_with(".js") {
+            // this URL is actually from a crate-internal path, serve it there instead
+            rendering_time.step("serve JS for crate");
+            let krate = spawn_blocking({
+                let crate_name = crate_name.clone();
+                let version = version.clone();
+                move || {
+                    let mut conn = pool.get()?;
+                    CrateDetails::new(&mut *conn, &crate_name, &version, &version, None)
+                }
+            })
+            .await?
+            .ok_or(AxumNope::ResourceNotFound)?;
+
+            match spawn_blocking({
```

Comment:
ideally we'd combine these two into the same async task; but I don't know if it matters in practice.

Reply:
In practice the difference is probably negligible, but so near to each other the optimization is easy to add.